pkg/registry/common/connect/nse_server.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connect
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/registry"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/sdk/pkg/registry/common/clienturl"
+	"github.com/networkservicemesh/sdk/pkg/registry/core/adapters"
+)
+
+type nseCacheEntry struct {
+	expiryTime time.Time
+	client     registry.NetworkServiceEndpointRegistryClient
+}
+
+type connectNSEServer struct {
+	dialOptions       []grpc.DialOption
+	clientFactory     func(ctx context.Context, cc grpc.ClientConnInterface) registry.NetworkServiceEndpointRegistryClient
+	cache             map[string]*nseCacheEntry
+	cacheLock         sync.Mutex
```

Comment:
Why global locks rather than a typed sync.Map?

Reply:
I've used executor for synchronization here because an executor can kill goroutine in the case when all events completed.

Reply:
But you still have a cacheLock mutex doing global locking... that's the part I don't understand.  Why should we have *any* global locks here... there are tools that allow us to avoid them.

Reply:
https://github.com/networkservicemesh/sdk/pull/374#discussion_r452300324